Improve tests

veerajsk opened the pull request.

veerajsk committed:
WIP

veerajsk committed:
Minor fixes

veerajsk committed:
Add sample test plans

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

veerajsk committed:
Modify transitions, gen code

veerajsk committed:
WIP

veerajsk committed:
way to set transitions

veerajsk committed:
First run without crash

veerajsk committed:
First working test

veerajsk committed:
Fix format

veerajsk committed:
Plan success test

veerajsk committed:
Minor change

veerajsk committed:
WIP

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

veerajsk committed:
Regenerate code

veerajsk committed:
Add isStateActive method

triccyx commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -100,7 +100,27 @@ class TestContext : public alica::AlicaContext
      */
     bool isPlanActive(int64_t id) const;
```
Change from `id` to `std::string name`

veerajsk replied:
didn't change this API because the other tests still use it. Ideally this API is not needed since getActivePlan should suffice

triccyx commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -100,7 +100,27 @@ class TestContext : public alica::AlicaContext
      */
```
`bool isStateActive(int64_t id) const;` can be removed or make private?

veerajsk replied:
yes, it can be removed. It is still there because other tests use it

veerajsk committed:
Multi plan instance test & fixes

veerajsk committed:
Fix relative diretory of behaviour:

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

triccyx commented on alica_test_utility/src/TestContext.cpp:
```diff
@@ -62,4 +64,235 @@ bool TestContext::isPlanActive(int64_t id) const
+    if (name.empty()) {
+        return nullptr;
+    }
+    return name[0] == '/' ? /* fully qualified name */ followRunningPlanPath(name) : /* just name of beh/plan */ searchRunningPlanTree(name);
```
I don't like  too much Elvis. But ok

veerajsk replied:
did not get the reference

triccyx replied:
I mean operator `?`  that is not standard c. Here we call it `the Elvis operator` 
However nothing important.

triccyx commented on supplementary/alica_ros1/alica_tests/src/test/Expr/src/ParallelSuccessOnCondPlan3288843407985944525.cpp:
```diff
@@ -0,0 +1,98 @@
+bool PreCondition2767999024419231358::evaluate(std::shared_ptr<RunningPlan> rp, const IAlicaWorldModel* wm)
+{
+    /*PROTECTED REGION ID(602472383731208509) ENABLED START*/
+    std::cout << "The PreCondition 2767999024419231358 in Transition '602472383731208509' is not implement yet!" << std::endl;
```
For some reason, std::cout on my system won't work. I use std::cerr

bjoernschroeder commented on alica_test_utility/src/TestContext.cpp:
```diff
@@ -62,4 +64,235 @@ bool TestContext::isPlanActive(int64_t id) const
+    if (results.empty() || results.size() > 1) {
+        return nullptr;
+    }
+    return results.front();
```
Can we add warnings if we cant find the correct node / find multiple with the same name? This could help with debugging failing tests

veerajsk replied:
can probably add a `getFailureInfo` method like the lbc mockups which returns the previous failure

bjoernschroeder commented on supplementary/alica_ros1/alica_tests/CMakeLists.txt:
```diff
@@ -46,36 +46,37 @@ if(BUILD_TESTING)
+        #src/test/test_inherit_blackboard.cpp
+        #src/test/test_failure_handling.cpp
+        #src/test/test_alica_logger.cpp
+        #src/test/test_legacy_conditions.cpp
```
have to be activated again

bjoernschroeder commented on supplementary/alica_ros1/alica_tests/src/test/Expr/src/Behaviour/SuccessOnInitBeh3821787310391665935.cpp:
```diff
@@ -0,0 +1,12 @@
+{
+// Behaviour:SuccessOnInitBeh
+
+} /* namespace alica */
```
What is this file doing? It looks like we can remove it from codegen in another PR

veerajsk replied:
why should this file be removed? It is the only file that implements the behaviour.

bjoernschroeder replied:
This seems to be the implementation file of the domain condition of the behaviour, not the behaviour itself. Also the file only contains the alica namespace without any implementation. But it will be removed when we remove codegen

bjoernschroeder commented on supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/Behaviour/SuccessOnInitBeh3821787310391665935.h:
```diff
@@ -0,0 +1,14 @@
+// Add additional includes here
+/*PROTECTED REGION END*/
+
+} /* namespace alica */
```
Maybe we can remove unused domain conditions from codegen for behaviours as well in a different PR

veerajsk replied:
yeah. Even better, we get rid of code gen entirely soon

bjoernschroeder commented on supplementary/alica_ros1/alica_tests/src/test/test_success.cpp:
```diff
@@ -0,0 +1,110 @@
+TEST_F(TestSuccessFixture, behSuccessTest)
+{
+    ASSERT_TRUE(_tc->setTransitionCond("TestMasterPlan", "ChooseTestState", "BehSuccessTestState"));
+    STEP_UNTIL2(_tc, _tc->getActiveBehaviour("SuccessOnInitBeh"));
```
You should be able to use STEP_UNTIL here, when you pass the context to the makro, it will automatically forward it to STEP_UNTIL2

veerajsk committed:
Reenable old tests

veerajsk committed:
Add failure reason

veerajsk committed:
Comments & print failure info

triccyx commented on supplementary/alica_ros1/alica_tests/src/test/test_success.cpp:
```diff
@@ -0,0 +1,110 @@
+namespace alica::test
+{
+
+class TestSuccessFixture : public ::testing::Test
```
Would you consider moving this Fixture to a separate header file to be reused in other tests?

veerajsk replied:
Yes, this is the idea. Its in this file for now until you refactor the tests to use the new APIs & design.

Dale-Koenig commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -90,17 +91,92 @@ class TestContext : public alica::AlicaContext
+     * blackboard & the TriggerFromInputCond condition to be used for the transition
+     * @param runningPlanName The name (fully qualified if required) of the plan in which the transition is present. This plan should be active
+     * @param inState The name of the incoming state of the transition
+     * @param outState The name of the outgoing state of the transition
```
Document return value for these.  I assume it is false just if it fails for some reason (states dont exist, or plan isn't running)?

veerajsk replied:
yes, & if the corresponding blackboard key for that transition does not exist

veerajsk replied:
I will add the documentation for it

Dale-Koenig commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -90,17 +91,92 @@ class TestContext : public alica::AlicaContext
+    /**
+     * Reset all transitions in the given plan, notes:
+     * The plan should be active
+     * Typically this method has to be called in all plans that use the TriggerFromInputCond condition in the init
```
Shouldn't such plans default to evaluating transitions to false when they init, and so we usually dont need to call this?

veerajsk replied:
I guess it should. unordered_map's emplace should value initialize default types to 0 & it does seem to be the case from my tests. So I will remove the calls to this method in the plan's init, but retain the method as a util function.

veerajsk replied:
I tried & it does not work at the moment because there is no support for types yet. So the type is std::any & the default constructor initialises it to an empty value. We can remove this when the types PR is merged. I'll add it back in for now

Dale-Koenig commented on supplementary/alica_ros1/alica_tests/src/test/test_success.cpp:
```diff
@@ -0,0 +1,110 @@
+    auto plan = _tc->getActivePlan("SuccessOnInitPlan");
+    ASSERT_NE(plan, nullptr) << _tc->getLastFailure();
+    STEP_UNTIL(_tc, _tc->isSuccess(plan));
+    ASSERT_TRUE(_tc->isSuccess(plan));
```
We should really introduce STEP_UNTIL_ASSERT_TRUE and STEP_UNTIL_ASSERT_EQ

veerajsk replied:
I think STEP_UNTIL_ASSERT_TRUE is sufficient since it can be used for all comparisons without having to add new STEP_UNTIL* macros for each type of comparison

Dale-Koenig replied:
At least in gbc I find EQ is used more frequently that TRUE .  Everything else is not so common

veerajsk replied:
ok, added it

Dale-Koenig commented on supplementary/alica_ros1/alica_tests/src/test/test_success.cpp:
```diff
@@ -0,0 +1,110 @@
+
+TEST_F(TestSuccessFixture, multiPlanInstanceSuccessTest)
+{
+    ASSERT_TRUE(_tc->setTransitionCond("TestMasterPlan", "ChooseTestState", "MultiPlanInstanceSuccessTestState")) << _tc->getLastFailure();
```
Comments+some organization.

veerajsk replied:
I'll add comments, but what organization are you referring to?

Dale-Koenig commented on supplementary/alica_ros1/alica_tests/src/test/test_success.cpp:
```diff
@@ -0,0 +1,110 @@
+    std::unique_ptr<ros::AsyncSpinner> _spinner;
+};
+
+TEST_F(TestSuccessFixture, behSuccessTest)
```
I prefer with each test you give a one or two line comment saying in human words what the test is trying to confirm

Dale-Koenig reviewed (commented):
What's with all the autogenerated files with transitions that aren't implemented?

veerajsk commented:
> What's with all the autogenerated files with transitions that aren't implemented?

@Dale-Koenig those transitions are for legacy transition condition support & therefore not implemented since the plans use the new transition conditions. I had wanted to put a deprecated comment there in the autogen code, but I suppose it got missed in the review comments & got merged

veerajsk committed:
Address review comments

- Add STEP_UNTIL_ASSERT_TRUE macro
- Add comments for the tests
- Add missing comment for the return value of setTransitionCond

veerajsk committed:
Review comment - Add step until assert eq

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

veerajsk committed:
Merge branch 'devel' into vsk_improve_tests

veerajsk committed:
Fixes

- Handle lvalues in set API correctly i.e. use std::decay before
determining the type (fixes compile error)
- Add tests for the above
- Specify types & default values for blackboard keys now that they are
supported
